src/org/labkey/test/tests/core/login/PasswordTest.java

```diff
@@ -253,13 +253,105 @@ public void testPasswordParameter()
         assertTrue("Expected email/password in URL to be rejected.", rejectedProperly);
     }
 
+    @Test
+    public void testChooseNewPasswordMessages() throws IOException
+    {
+        // Hold an admin connection open, allowing us to reset the config regardless of what happens during the test
+        Connection adminConnection = WebTestHelper.getRemoteApiConnection();
+        DbLoginProperties savedProperties = DbLoginUtils.getDbLoginConfig(adminConnection);
+
+        try
+        {
+            DbLoginUtils.setDbLoginConfig(adminConnection,
+                PasswordStrength.Good,
+                PasswordExpiration.Never
+            );
+
+            // Set a weak password
+            signOut();
+            String resetUrl = userInitiatePasswordReset(USER);
+            beginAt(resetUrl);
+            new SetPasswordForm(getDriver())
+                .setNewPassword(WEAK_PASSWORD)
+                .clickSubmit();
+
+            // Test bogus password first
+            signOut();
+            signInShouldFail("bogus", "The email address and password you entered did not match any accounts on file.");
+            signIn(USER, WEAK_PASSWORD);
+            signOut();
+
+            // Change the configuration and test password that no longer meets complexity requirements
+            DbLoginUtils.setDbLoginConfig(adminConnection,
+                PasswordStrength.Strong,
+                PasswordExpiration.Never
+            );
+            signInShouldFail(WEAK_PASSWORD, "Your password does not meet the complexity requirements; please choose a new password.");
+            String strongPassword = VERY_STRONG_PASSWORD + "!";
+            changeInvalidPassword(WEAK_PASSWORD, strongPassword);
+
+            // Change the configuration and test expired password
+            DbLoginUtils.setDbLoginConfig(adminConnection,
+                PasswordStrength.Strong,
+                PasswordExpiration.FiveSeconds
+            );
+            // Wait five seconds for expiration
+            sleep(5000);
```

Comment:
If 5 seconds is the amount of time the system needs to expire a password, maybe the test should wait 6 seconds to give a little buffer.